modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ProjectedTuple.java

```diff
@@ -0,0 +1,322 @@
+    }
+
+    @Override
+    public UUID uuidValue(int col) {
+        return delegate.uuidValue(projection[col]);
+    }
+
+    @Override
+    public BitSet bitmaskValue(int col) {
+        return delegate.bitmaskValue(projection[col]);
+    }
+
+    @Override
+    public LocalDate dateValue(int col) {
+        return delegate.dateValue(projection[col]);
+    }
+
+    @Override
+    public LocalTime timeValue(int col) {
+        return delegate.timeValue(projection[col]);
+    }
+
+    @Override
+    public LocalDateTime dateTimeValue(int col) {
+        return delegate.dateTimeValue(projection[col]);
+    }
+
+    @Override
+    public Instant timestampValue(int col) {
+        return delegate.timestampValue(projection[col]);
+    }
+
+    @Override
+    public ByteBuffer byteBuffer() {
+        normalizeIfNeeded();
+
+        return delegate.byteBuffer();
+    }
+
+    private void normalizeIfNeeded() {
+        if (normalized) {
+            return;
+        }
+
+        if (schema != null) {
+            normalizeSlow();
+        } else {
+            normalizeFast();
+        }
+    }
+
+    private void normalizeSlow() {
+        assert schema != null;
+
+        var builder = new BinaryTupleBuilder(projection.length);
+        var newProjection = new int[projection.length];
+
+        for (int i = 0; i < projection.length; i++) {
+            int col = projection[i];
+
+            newProjection[i] = i;
+
+            Element element = schema.element(col);
+
+            BinaryRowConverter.appendValue(builder, element, schema.value(delegate, col));
+        }
+
+        delegate = new BinaryTuple(projection.length, builder.build());
+        projection = newProjection;
+        normalized = true;
+    }
+
+    private void normalizeFast() {
+        var newProjection = new int[projection.length];
+        ByteBuffer tupleBuffer = delegate.byteBuffer();
+        int[] requiredColumns = projection;
+
+        var parser = new BinaryTupleParser(delegate.elementCount(), tupleBuffer);
+
+        // Estimate total data size.
+        var stats = new Sink() {
+            int estimatedValueSize = 0;
+
+            @Override
+            public void nextElement(int index, int begin, int end) {
+                estimatedValueSize += end - begin;
+            }
+        };
+
+        for (int columnIndex : requiredColumns) {
+            parser.fetch(columnIndex, stats);
+        }
+
+        // Now compose the tuple.
+        BinaryTupleBuilder builder = new BinaryTupleBuilder(requiredColumns.length, stats.estimatedValueSize);
+
+        int pos = 0;
+
+        for (int columnIndex : requiredColumns) {
+            parser.fetch(columnIndex, (index, begin, end) -> {
+                if (begin == end) {
+                    builder.appendNull();
+                } else {
+                    builder.appendElementBytes(tupleBuffer, begin, end - begin);
+                }
+            });
+
+            newProjection[pos++] = columnIndex;
+        }
```

Comment:
I suggest to remove `pos` counter.

```suggestion
                for (int i = 0; i < requiredColumns.length; i++) {
            parser.fetch(requiredColumns[i], (index, begin, end) -> {
                if (begin == end) {
                    builder.appendNull();
                } else {
                    builder.appendElementBytes(tupleBuffer, begin, end - begin);
                }
            });

            newProjection[i] = requiredColumns[i];
        }
```

Reply:
fixed